Windows 兼容性 bug：subprocess 缺少 shell=True 及参数格式错误

## 问题描述

在 Windows 10 (Git Bash) 环境下运行 `update.py` 时遇到两个 bug：

### Bug 1: `get_npm_root()` 和 `get_version_npm()` 缺少 `shell=True`

Windows 上 npm 是 `.cmd` 批处理文件，`subprocess.run()` 必须加 `shell=True` 才能正确找到并执行。

**影响**：`get_npm_root()` 返回 `None`，脚本报 `Could not determine npm global root` 后退出。

**复现**：
```
Current version: 2.1.116
Latest version: 2.1.119
[ERROR] Could not determine npm global root
```

**修复**（`get_npm_root`，约第 209 行）：
```python
def get_npm_root() -> Optional[Path]:
    try:
        use_shell = platform.system() == 'Windows'
        result = subprocess.run(
            ['npm', 'root', '-g'],
            capture_output=True, text=True, timeout=5,
            shell=use_shell
        )
```

`get_version_npm`（约第 143 行）同理加 `shell=use_shell`。

### Bug 2: `do_install()` 中 `Popen` + `shell=True` + 列表参数导致 `FileNotFoundError`

Windows 上 `shell=True` 时，`Popen` 的 `args` 参数应为字符串，而非列表。传列表会导致 `CreateProcess` 找不到可执行文件。

**影响**：
```
FileNotFoundError: [WinError 2] 系统找不到指定的文件。
```

**修复建议**（`do_install`，约第 343 行）：
```python
def do_install(target_version: str, npm_root: Path) -> bool:
    use_shell = platform.system() == 'Windows'
    if use_shell:
        install_cmd = f'npm install -g --force @anthropic-ai/claude-code@{target_version}'
    else:
        install_cmd = ['npm', 'install', '-g', '--force', f'@anthropic-ai/claude-code@{target_version}']

    # ...
    proc = subprocess.Popen(
        install_cmd,
        stdout=log_fd,
        stderr=subprocess.STDOUT,
        text=True,
        shell=use_shell,
    )
```

### Bug 2 的连锁问题：安装失败未恢复重命名的 exe

当 `do_install` 因 `FileNotFoundError` 崩溃时，`install_with_retry` 的异常未被捕获（直接抛到 `main`），导致 `restore_renamed_exes()` 不会执行。结果是 `claude.exe` 保持为 `claude.exe.old`，后续运行报 `Claude Code not found`。

**建议**：在 `do_install` 中加 try/except 包裹 `Popen`，或在 `install_with_retry` 的调用外增加异常保护，确保失败时恢复重命名的文件。

## 环境信息
- Windows 10 Pro 10.0.19045
- Git Bash (MSYS2)
- Python 3.13 (Microsoft Store 版)
- npm 11.12.1
- Node.js v22.14.0

--- a/update-claude-code/scripts/update.py
+++ b/update-claude-code/scripts/update.py
@@ -57,6 +57,8 @@ LOCKED_EXE_RELPATHS = [
     Path('@anthropic-ai') / 'claude-code' / 'node_modules' / '@anthropic-ai' / 'claude-code-win32-x64' / 'claude.exe',
 ]
 
+IS_WINDOWS = platform.system() == 'Windows'
+
 
 def extract_version(text: str) -> Optional[str]:
     match = VERSION_PATTERN.search(text)
@@ -117,13 +119,12 @@ def log_error(message):
 
 def get_current_version() -> Optional[str]:
     try:
-        use_shell = platform.system() == 'Windows'
         result = subprocess.run(
             ['claude', '--version'],
             capture_output=True,
             text=True,
             timeout=5,
-            shell=use_shell
+            shell=IS_WINDOWS
         )
         if result.returncode == 0:
             return extract_version(result.stdout + result.stderr)
@@ -146,7 +147,8 @@ def get_version_npm() -> Optional[str]:
             ['npm', 'view', '@anthropic-ai/claude-code', 'version'],
             capture_output=True,
             text=True,
-            timeout=15
+            timeout=15,
+            shell=IS_WINDOWS
         )
         if result.returncode == 0:
             version = result.stdout.strip()
@@ -210,7 +212,8 @@ def get_npm_root() -> Optional[Path]:
     try:
         result = subprocess.run(
             ['npm', 'root', '-g'],
-            capture_output=True, text=True, timeout=5
+            capture_output=True, text=True, timeout=5,
+            shell=IS_WINDOWS
         )
         if result.returncode == 0:
             return Path(result.stdout.strip())
@@ -336,23 +339,25 @@ def cleanup_npm_temp_dirs(npm_root: Path):
 # =============================================================================
 
 def do_install(target_version: str, npm_root: Path) -> bool:
-    install_cmd = ['npm', 'install', '-g', '--force', f'@anthropic-ai/claude-code@{target_version}']
+    if IS_WINDOWS:
+        install_cmd = f'npm install -g --force @anthropic-ai/claude-code@{target_version}'
+    else:
+        install_cmd = ['npm', 'install', '-g', '--force', f'@anthropic-ai/claude-code@{target_version}']
     log_info(f"Installing version {target_version} (timeout: {INSTALL_TIMEOUT}s)...")
 
     log_path = os.path.join(tempfile.gettempdir(), f'claude-update-{os.getpid()}.log')
 
     log_fd = open(log_path, 'w')
-    proc = subprocess.Popen(
-        install_cmd,
-        stdout=log_fd,
-        stderr=subprocess.STDOUT,
-        text=True,
-    )
-
     success = False
     try:
+        proc = subprocess.Popen(
+            install_cmd,
+            stdout=log_fd,
+            stderr=subprocess.STDOUT,
+            text=True,
+            shell=IS_WINDOWS,
+        )
         proc.wait(timeout=INSTALL_TIMEOUT)
-
         if proc.returncode == 0:
             success = True
         else:
@@ -361,12 +366,13 @@ def do_install(target_version: str, npm_root: Path) -> bool:
         log_error(f"Install timed out after {INSTALL_TIMEOUT}s")
         proc.kill()
         proc.wait()
-
-    log_fd.close()
-
-    if not success:
-        _print_log_tail(log_path, 10)
-    _safe_unlink(log_path)
+    except Exception as e:
+        log_error(f"Install error: {e}")
+    finally:
+        log_fd.close()
+        if not success:
+            _print_log_tail(log_path, 10)
+        _safe_unlink(log_path)
     return success
 
 
@@ -393,7 +399,7 @@ def install_with_retry(target_version: str, npm_root: Path) -> bool:
         cleanup_npm_temp_dirs(npm_root)
 
         renamed = []
-        if platform.system() == 'Windows':
+        if IS_WINDOWS:
             cleanup_old_files(npm_root)
             renamed = rename_locked_exes(npm_root)
 
@@ -403,7 +409,7 @@ def install_with_retry(target_version: str, npm_root: Path) -> bool:
 
         if success:
             # Clean up .old files after successful install
-            if platform.system() == 'Windows':
+            if IS_WINDOWS:
                 cleanup_old_files(npm_root)
             return True
 

